### ISSUE TITLE
Why LayerNorm layers are frozen?

### ISSUE THREAD

**Issue by iliaschalkidis** [hosein-m/TF-Adapter-BERT#1]
Hi @hosein-m,

I read your code and the paper (https://arxiv.org/pdf/1902.00751.pdf). According to the paper, the  LayerNorm layers shall be trainable. Am I missing something?

https://github.com/hosein-m/TF-Adapter-BERT/blob/8ddad140dc8c61b5db4db50d47fc258b0e9868cb/run_tf_glue_adapter_bert.py#L110

**Comment by hmohebbi** [comment]
Hi @iliaschalkidis, thank you for the correction. I agree with you. As I remember, fixing the LayerNorm layers' parameters yielded a better GLUE performance score in my experiments.

**Comment by iliaschalkidis** [comment]
Great, thanks!

I would also recommend you to let the `pooler` trainable, too. It's a randomly initialized layer that shall be fine-tuned, otherwise, it will remain as it is (randomly initialized). You could also skip this layer using `bert(inputs)[:,0]`, the final `CLS`token, as a document representation instead of the `pooler`, `bert(inputs)[1]`  one.

I also was wondering if the `TruncatedNormal` initialization with `stddev=0.02` is aligned with the paper part:

"_With the skip-connection, if the parameters of the projection layers are initialized to near-zero, the module is initialized to an approximate identity function._"

In the official implementation the `stddev=1e-3`: https://github.com/google-research/adapter-bert/blob/1a31fc6e92b1b89a6530f48eb0f9e1f04cc4b750/modeling.py#L321, which seems to better approximate _"near-zero"_, as BERT weights are anyway very close to zero already. 😄 


Thanks for your implementation!

**Comment by hmohebbi** [comment]
Thanks for the detailed comments. @iliaschalkidis 

It is worth mentioning that after calling the `.from_pretrained` method in line [74](https://github.com/hosein-m/TF-Adapter-BERT/blob/8ddad140dc8c61b5db4db50d47fc258b0e9868cb/run_tf_glue_adapter_bert.py#L74), the `pooler` layer will be initialized with its pre-trained weights which were trained by the NSP task. (see https://github.com/huggingface/transformers/issues/300)

**Comment by iliaschalkidis** [comment]
Sorry @hosein-m, one last question:

https://github.com/hosein-m/TF-Adapter-BERT/blob/8ddad140dc8c61b5db4db50d47fc258b0e9868cb/modeling_tf_adapter_bert.py#L10

Does this mean that you use the very same `AdapterModule`, for 2 layers, back-to-back?

**Comment by hmohebbi** [comment]
Sorry @iliaschalkidis for closing the issue!

According to Houlsby’s architecture, there must be two Adapter modules in each Transformer layer: one in the `TFBertSelfOutput` component and the other in the `TFBertOutput`, which both of them must have a shared weight. So, this line fulfills this purpose and aligned with this part of the paper:
"_In each layer, the total number of parameters added per layer, including biases, is 2md + d + m._"

Please let me know if I miss something :)

**Comment by iliaschalkidis** [comment]
> Sorry @iliaschalkidis for closing the issue!
> 
> According to Houlsby’s architecture, there must be two Adapter modules in each Transformer layer: one in the `TFBertSelfOutput` component and the other in the `TFBertOutput`, which both of them must have a shared weight. So, this line fulfills this purpose and aligned with this part of the paper:
> "_In each layer, the total number of parameters added per layer, including biases, is 2md + d + m._"
> 
> Please let me know if I miss something :)

Yeah, I show this line in the article and suspected this is your motivation. They should have phrased this better and clearer, like _"The two adapter layers are tied."_ or something similar. I cannot validate this in the original implementation.

Thanks again!
